### PR TITLE
IDEMPIERE-2853: Support for IADTabPanel as factory to add custom Tab …

### PR DESCRIPTION
**Pull request by dpansheriya**
…types

sh: q: command not found

I closed original PR https://github.com/idempiere/idempiere/pull/474 and created new PR

### CODE REVIEW

File: org.adempiere.base/src/org/compiere/model/GridTab.java
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/GridTab.java:593]: please add javadoc

File: org.adempiere.base/src/org/compiere/model/GridTabVO.java
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/GridTabVO.java:550]: please add javadoc

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ADSortTab.java
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ADSortTab.java:85]: what's the use of this new constructor ?

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/AbstractADWindowContent.java
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/AbstractADWindowContent.java:650]: add new Exception(errormsg) parameter to get stacktrace into log would help debugging.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/AbstractADWindowContent.java
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/AbstractADWindowContent.java:1136]: shouldn't this be replaced by the new isEnableCustomizeButton() method ?

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/AbstractADWindowContent.java
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/AbstractADWindowContent.java:1285]: same here, should be replaced by the new isEnableProcessButton() method

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/IADTabpanel.java
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/IADTabpanel.java:39]: why the need of gridWindow parameter here when you have gridTab.getGridWindow() ?

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/factory/IADTabPanelFactory.java
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/factory/IADTabPanelFactory.java:23]: please add javadoc

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/factory/DefaultTabPanelFactory.java
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/factory/DefaultTabPanelFactory.java:29]: should use constant from model class instead

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ADTabpanel.java
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ADTabpanel.java:2136]: this is a good candidate to move to interface as default method.

**Comment by hengsin** [comment]
Need to regenerate model class for ad_tab

**Comment by hengsin** [comment]
also, it would be good to add unit test to verify the changes doesn't break existing loading of grid window and grid tab.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ADSortTab.java
Comment (dpansheriya) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ADSortTab.java:85]: It used when we get object for osgi service in Extensions.getADTabPanel. We later call init method after retrieving object of implementer class.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/AbstractADWindowContent.java
Comment (dpansheriya) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/AbstractADWindowContent.java:1136]: Agree. Changing.

**Review by CarlosRuiz-globalqss — COMMENTED** [review]
more comments on ticket

**Comment by dpansheriya** [comment]
I am working on integrating all suggestion by @CarlosRuiz-globalqss and @hengsin. mean while want one more clarification.
FYI.. For backward compellability, we are still supporting isSortTab flag. do you think we should add migration script to use tab type and drop isSortTab flag?

**Comment by hengsin** [comment]
hi @dpansheriya, to integrate this to core, we must migrate from issorttab to tabtype. having both in UI and dictionary is just asking for confusion and trouble.

**Comment by dpansheriya** [comment]
> hi @dpansheriya, to integrate this to core, we must migrate from issorttab to tabtype. having both in UI and dictionary is just asking for confusion and trouble.

@hengsin Agree. But I suggest that we mark isSortTab field in-active at moment and later we drop it to avoid any in progress PR to fail. Or do you agree that all next commit must be corrected to use new way of specifying tab type and let all contributor to fix their PR once this feature is in master?

**Comment by hengsin** [comment]
this pr must inactivate issorttab and have migration script to migrate that to tabtype. after this have been merged to core, subsequent pr must use the tabtype instead of issorttab. lets keep it simple.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ADTabpanel.java
Comment (dpansheriya) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ADTabpanel.java:2136]: @hengsin I did not understand this suggestion. isEnableProcessButton, isEnableCustomizeButton are already in IADTabpanel interface.

Comment (hengsin) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ADTabpanel.java:2136]: Since it is unlikely for implementation class to change the "return getToolbarButtons().size() > 0 && !gridTab.isNew();" logic, instead of making it a must for every implementation of IADTabpanel interface to implement the method, change it to public default boolean isEnableProcessButton() {return getToolbarButtons().size() > 0 && !getGridTab().isNew();} at IADTabpanel interface.

File: org.adempiere.base/src/org/compiere/model/GridTab.java
Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/GridTab.java:597]: Done

File: org.adempiere.base/src/org/compiere/model/GridTabVO.java
Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/GridTabVO.java:550]: Done

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/AbstractADWindowContent.java
Comment (dpansheriya) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/AbstractADWindowContent.java]: Done

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/AbstractADWindowContent.java
Comment (dpansheriya) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/AbstractADWindowContent.java]: Done

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/AbstractADWindowContent.java
Comment (dpansheriya) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/AbstractADWindowContent.java]: Done at all placess.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/IADTabpanel.java
Comment (dpansheriya) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/IADTabpanel.java]: dropped WindowNo and gridWindow from argument.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/factory/IADTabPanelFactory.java
Comment (dpansheriya) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/factory/IADTabPanelFactory.java:29]: Done

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/factory/DefaultTabPanelFactory.java
Comment (dpansheriya) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/factory/DefaultTabPanelFactory.java]: done

File: migration/i8.2z/oracle/201601221200_IDEMPIERE-2853.sql
Comment (dpansheriya) [reply migration/i8.2z/oracle/201601221200_IDEMPIERE-2853.sql]: Fixed oracle seqNo to 400. Only one deviations all other were matching.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ADTabpanel.java
Comment (dpansheriya) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ADTabpanel.java]: In latest commit, I made this change. Please verify that I have done correctly.

File: org.adempiere.base/src/org/compiere/model/X_AD_Tab.java
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/X_AD_Tab.java:329]: looks kind of strange to me that the list name and constant name is completely different. I means it looks a lot more natural if it is AD_TABTYPE_Form="FORM" and AD_TABTYPE_SortTab="SORT"

Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/X_AD_Tab.java:329]: @hengsin originally the isSortTab column is named as Order Tab, So we followed same conventions. Do you think it is good to fix and change label to to Sort?

Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/X_AD_Tab.java]: @hengsin this fix committed in comitt https://github.com/idempiere/idempiere/pull/579/commits/7eb84b31564af975ef7a6aad48fcbbf5e1b07c6a
Also resolved merge conflict with Master branch.

**Comment by hengsin** [comment]
we should keep this on hold until we can agree upon a design that's inclusive of IDEMPIERE-4726

**Comment by CarlosRuiz-globalqss** [comment]
Hi @dpansheriya - I'm interested in adding this to the new server test-feature, but the pull request has recent conflicts.

Can you please check the conflicts and I'll install this in test-feature for community tests?

**Comment by dpansheriya** [comment]
> Hi @dpansheriya - I'm interested in adding this to the new server test-feature, but the pull request has recent conflicts.
> 
> Can you please check the conflicts and I'll install this in test-feature for community tests?

@CarlosRuiz-globalqss Let me merge it again to latest code.

**Comment by dpansheriya** [comment]
@CarlosRuiz-globalqss I merged branch and resolved conflicts.

**Comment by CarlosRuiz-globalqss** [comment]
ready @dpansheriya - this feature is installed now on this server:
https://test-feature.idempiere.org:3243/webui/index.zul

in case you want to install a sample for community to help with tests

**Comment by dpansheriya** [comment]
Thanks Carlos, will try to install Calendar plugin by Tuesday.

**Comment by dpansheriya** [comment]
@CarlosRuiz-globalqss I deployed Calendar Tab plugin on feature test server and configurated Tab named "Request View" on Business Partner window. This tab shows only request assigned to Parent Sales Representative or Request created for Parent BPartner.

**Comment by CarlosRuiz-globalqss** [comment]
Hi @dpansheriya - I'm revisiting this but it seems the server test-feature is down  :-(

The old example plugin you provided com.adaxa.calendarTab doesn't work with this new version.  Can you please provide (upload to the ticket maybe) a working example.

**Comment by dpansheriya** [comment]
> Hi @dpansheriya - I'm revisiting this but it seems the server test-feature is down :-(
> 
> The old example plugin you provided com.adaxa.calendarTab doesn't work with this new version. Can you please provide (upload to the ticket maybe) a working example.

 I uploaded plugin on Jira ticket. test-feature server is down so not able to install there.

**Comment by dpansheriya** [comment]
@CarlosRuiz-globalqss I also converted mercurial repository to git and uploaded https://github.com/logilite/CalendarTab. Please note that still mavenization is pening.

**Review by CarlosRuiz-globalqss — APPROVED** [review]
Tested successfully

**Pull request merged**
